agents/subtask_predictor.py

```diff
@@ -0,0 +1,311 @@
+from tqdm import tqdm
+import time
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.utils.data import DataLoader
+from torch.distributions.categorical import Categorical
+import wandb
+import tensorflow as tf
+import torch
+import random
+def get_output_shape(model, image_dim):
+    return model(th.rand(*(image_dim))).data.shape[1:]
+
+def weights_init_(m):
+    if hasattr(m, 'weight') and m.weight is not None and len(m.weight.shape) > 2:
+        th.nn.init.xavier_uniform_(m.weight, gain=1)
+    if hasattr(m, 'bias') and m.bias is not None and isinstance(m.bias, th.Tensor):
+        th.nn.init.constant_(m.bias, 0)
+
+def merge_two_dicts(x, y):
+    z = x.copy()   # start with keys and values of x
+    z.update(y)    # modifies z with keys and values of y
+    return z
+
+class SubtaskPredictor(nn.Module):
+    def __init__(self, device, visual_obs_shape, agent_obs_shape, depth=16, act=nn.ReLU, hidden_dim1=256, hidden_dim2=512, hidden_dim3 = 1024, hidden_dim4=4000):
+        super(SubtaskPredictor, self).__init__()
+        self.device = device
+        self.act = act
+        self.hidden_dim1 = hidden_dim1
+        self.hidden_dim2 = hidden_dim2
+        self.hidden_dim3 = hidden_dim3
+        self.hidden_dim4 = hidden_dim4
+        self.use_visual_obs = np.prod(visual_obs_shape) > 0
+        assert len(agent_obs_shape) == 1
+        self.use_agent_obs = np.prod(agent_obs_shape) > 0
+        self.subtasks_obs = Subtasks.NUM_SUBTASKS 
+    
+        # Define CNN for grid-like observations
+        if self.use_visual_obs:
+            self.kernels = (4, 4, 4, 4) if max(visual_obs_shape) > 64 else (3, 3, 3, 3)
+            self.strides = (2, 2, 2, 2) if max(visual_obs_shape) > 64 else (1, 1, 1, 1)
+            self.padding = (1, 1)
+            layers = []
+            current_channels = visual_obs_shape[0]
+            for i, (k, s) in enumerate(zip(self.kernels, self.strides)):
+                layers.append(nn.Conv2d(current_channels, depth, k, stride=s, padding=self.padding))
+                layers.append(nn.GroupNorm(1, depth))
+                layers.append(self.act())
+                current_channels = depth
+                depth *= 2
+
+            layers.append(nn.Flatten())
+            self.cnn = nn.Sequential(*layers)
+            self.cnn_output_shape = get_output_shape(self.cnn, [1, *visual_obs_shape])[0]
+            self.pre_flatten_shape = get_output_shape(self.cnn[:-1], [1, *visual_obs_shape])
+        else:
+            self.cnn_output_shape = 0
+
+        print("input size:", self.cnn_output_shape + agent_obs_shape[0] +self.subtasks_obs+ 10*self.subtasks_obs+ Subtasks.NUM_SUBTASKS)
+        # Define layers
+        self.rnn = nn.Sequential(
+            nn.Linear(self.cnn_output_shape + agent_obs_shape[0] +self.subtasks_obs+ 10*self.subtasks_obs+ Subtasks.NUM_SUBTASKS, self.hidden_dim4),
+            nn.Dropout(),
+            act()
+        )
```

Comment:
This isn't actually an rnn. An rnn takes its previous output plus a new input, but no where do I see you pass in the previous output (not that this needs to be an rnn, but either change the name or make it a proper rnn). To note, pytorch has ready-made rnns. Also, in most cases it is better to use an LSTM or a GRU rather than a plain rnn. LSTMs/GRUs contain gating mechanisms that help mitigate the vanishing/exploding gradient problem.